### ISSUE TITLE
Feature Request: Option to force 2FA at account signup in order to reduce spam

### ISSUE THREAD

**Issue by russsaidwords** [tootsuite/mastodon#13013]
The idea is that having 2FA enforced for signups would cut down on low effort spam account creation. I'm sure that this would slow a normal user down as well, but I'm willing to take that risk for my own instance. 

Six out of the eight most recent accounts created on my instance were actively or passively being used for spam - and this is a small instance, I can't even imagine the amount of administration that would need to happen if it were more popular. So my request is really an attempt to address that problem and not necessarily to gain more security for individual user accounts.

Thanks for all the hard work on Mastodon. See you in the Fediverse.

**Comment by Shleeble** [comment]
hell.. I'd be happy with some kind of once off SMS based MFA

**Comment by trwnh** [comment]
>hell.. I'd be happy with some kind of once off SMS based MFA

Forced collection of phone numbers is a bad idea, as not everyone has a phone number. Setting up TOTP can be done on any device, not just a phone, and nothing to do with a phone number.

**Comment by Shleeble** [comment]
> > hell.. I'd be happy with some kind of once off SMS based MFA
> 
> Forced collection of phone numbers is a bad idea, as not everyone has a phone number. Setting up TOTP can be done on any device, not just a phone, and nothing to do with a phone number.

Yeah.. Isn't that just enforcing TOTP MFA? I mean that's a great idea but MFA isn't exactly easy for non technical people.

What's some kind of single use token that's ubiquitous?

People don't like recaptcha.... Is there any popular local captcha we could use for signups?

**Comment by russsaidwords** [comment]
> MFA isn't exactly easy for non technical people.

It's not easy for bots, either. Is it possible? I'm sure it is. Is it worth it to a spammer to set up MFA for every account they want to open? Probably not. From what I've seen these are mostly low effort attempts to spam others / open an account for spamming others / increase SEO through linking back to spam websites. My hope is they just move on to an easier target and that the *people* that wind up setting up an account are more secure for the trouble. CAPTCHA secures only one side of the equation, and not that well. Having both would be grand, sure.

**Comment by Shleeble** [comment]
This is the armsrace thing.. if it's only a few more lines of code to generate the totp, then store the session token. 

All you've done is kicked out the older spambots.

**Comment by Shleeble** [comment]
This has the benefit of being able to see spambot ip ranges.

**Comment by russsaidwords** [comment]
It is an arms race, but you can't win an arms race through inaction. It's frustrating to run an instance that is open to public account creation and needing to remove spam accounts every few days. It's starting to feel like a job, and when that happens I tend to look for an automated solution. So really, I'm open to anything that cuts down on automated / spam account creation. Tying the account to a phone number or email (even if it's just a one time contact) would be fine by me. I'm open to whatever would be easiest to achieve the goal.

**Comment by Shleeble** [comment]
1) This also runs into the idea of trust between instances... hardcore spammers could create a fake activitypub server on as many domains as possible and just send as much spam as they want.

The shelve life is as long as it takes me to investigate and decide to block the domain name (subdomains are blocked automaticly).

2) I've argued that I wanted the ability to promote trusted "sister" instances that I trust is run by people in good faith... but that only makes sense if you're detecting inbound spam on non-trusted instances more than trusted instances.

I'd love to see something similar to us promoting hashtags as admins... "new instance detected. Do you want to trust it? and as I said before about trusted instances (new instance X detected - it is trusted by 6 other trusted instances)"

**Comment by Shleeble** [comment]
This kind of thing is helped by the OCAP design which is still in concept design.

**Comment by Gargron** [comment]
> This kind of thing is helped by the OCAP design which is still in concept design.

How is OCAP going to help spam sign-ups?

**Comment by Shleeble** [comment]
> > This kind of thing is helped by the OCAP design which is still in concept design.
> 
> How is OCAP going to help spam sign-ups?
Sorry, I didn't mean it'll solve spam - more of an example of a solution of a security/trust issue that's being built outside of the standard. (I should have flipped my two lines around to make sense)

**Comment by Shleeble** [comment]
To be back on topic... I know it's not a pretty solution, but for spam signups on our instances.
but having some kind of recaptcha or questionnaire is a classic solution.

In regards to questionnaires, If every instance instance has a different custom question - that means spammers need to waste their time to hardcode your instances question/answer pair

But those things are generally not translated for multiple languages. so I couldn't make a question like "What's the island state of Australia?" without likely losing a lot of non-english speakers.

**Comment by Shleeble** [comment]
There is a lot of previous work on this topic
https://www.w3.org/WAI/GL/wiki/Captcha_Alternatives_and_thoughts
https://www.w3.org/TR/turingtest/

Possible investigations:
https://www.hcaptcha.com/
https://github.com/subwindow/negative-captcha

**Comment by SebastianBerlin** [comment]
Any updates on this?

**Comment by hinricht** [comment]
I'd like to stay on topic, this issue is about *2FA* to fight against spam, and suggest that other solutions (like CAPTCHAs etc) should get discussed in a separate issue.
I'd relly like to enforce 2fa for all accounts on our server, not just to prevent registration spam but also to increase security.

**Comment by hex-m** [comment]
I don't think TOTP/WebAuthn would stop spammers (for long), but I would like to force my users to enable 2FA / MFA nonetheless.
